src/workerd/server/alarm-scheduler.h

```diff
@@ -0,0 +1,131 @@
+};
+
+inline uint KJ_HASHCODE(const ActorKey& k) { return kj::hashCode(k.uniqueKey, k.actorId); }
+
+class AlarmScheduler final : kj::TaskSet::ErrorHandler {
+  // Allows scheduling alarm executions at specific times, returning a promise representing
+  // the completion of the alarm event.
+public:
+  static constexpr auto RETRY_START_SECONDS = WorkerInterface::ALARM_RETRY_START_SECONDS;
+
+  static constexpr auto RETRY_MAX_TRIES = WorkerInterface::ALARM_RETRY_MAX_TRIES;
+  // Max number of "valid" retry attempts, i.e the worker returned an error
+
+  static constexpr auto RETRY_BACKOFF_MAX = 9;
+  // Bound for exponential backoff when RETRY_MAX_TRIES is exceeded due to internal errors.
+  // 2 << 9 is 1024 seconds, about 17 minutes. Total time spent in retries once the backoff limit
+  // is reached is over 30 minutes.
+
+  static constexpr auto RETRY_JITTER_FACTOR = 0.25;
+  // How much jitter should be applied to retry times to avoid bundled retries overloading
+  // some common dependency between a set of failed alarms
+
+  using GetActorFn = kj::Function<kj::Own<WorkerInterface>(kj::String)>;
+
+  AlarmScheduler(
+    const kj::Clock& clock,
+    kj::Timer& timer,
+    const SqliteDatabase::Vfs& vfs,
+    kj::PathPtr path);
+
+  kj::Date getAlarm(ActorKey actor);
+  bool setAlarm(ActorKey actor, kj::Date scheduledTime);
+  bool deleteAlarm(ActorKey actor);
+
+  void ensureInitialized();
+
+  void registerNamespace(kj::StringPtr uniqueKey, GetActorFn getActor);
+
+private:
+  const kj::Clock& clock;
+  kj::Timer& timer;
+  std::default_random_engine random;
+
+  struct Namespace {
+    GetActorFn getActor;
+  };
+  kj::HashMap<kj::StringPtr, Namespace> namespaces;
+  SqliteDatabase db;
+  kj::TaskSet tasks;
+
+  struct ScheduledAlarm {
+    kj::Own<ActorKey> actor;
+    kj::Date scheduledTime;
+    kj::Promise<void> task;
+    kj::Maybe<kj::Date> queuedAlarm = nullptr;
+    // Once started, an alarm can have a single alarm queued behind it.
+    bool started = false;
+
+    bool previousRetryCountedAgainstLimit = false;
+
+    // Counter for calculating backoff -- separate from retry, so we can reset backoff without losing
+    // the total count of retry attempts
+    uint32_t backoff = 0;
+
+    // Counter for retry attempts, whether or not they apply to the limit
+    uint32_t retry = 0;
+
+    // Counter for retry attempts that apply to the retry limit.
+    uint32_t countedRetry = 0;
+  };
+
+  kj::HashMap<ActorKey, ScheduledAlarm> alarms;
+
+  struct RetryInfo {
+    bool retry;
+    bool retryCountsAgainstLimit;
+  };
+  kj::Promise<RetryInfo> runAlarm(const ActorKey& actor, kj::Date scheduledTime);
+
+  void setAlarmInMemory(kj::Own<ActorKey> actor, kj::Date scheduledTime);
+
+  ScheduledAlarm scheduleAlarm(kj::Date now, kj::Own<ActorKey> actor, kj::Date scheduledTime);
+
+  kj::Promise<void> makeAlarmTask(kj::Duration delay, const ActorKey& actor, kj::Date scheduledTime);
+
+  SqliteDatabase::Statement stmtGetAlarm = db.prepare(R"(
```

Comment:
Can we make these maybes and initialize these inline to the functions that use them? I'd like for devs to be able to compare them against the indexes in use there.